core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -597,24 +597,45 @@ public PlanWithProperties visitSimpleTableExecuteNode(SimpleTableExecuteNode nod
         @Override
         public PlanWithProperties visitTableWriter(TableWriterNode node, StreamPreferredProperties parentPreferences)
         {
-            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences, node.getTarget());
+            return visitTableWriter(
+                    node,
+                    node.getPartitioningScheme(),
+                    node.getPreferredPartitioningScheme(),
+                    node.getSource(),
+                    parentPreferences,
+                    node.getTarget());
         }
 
         @Override
         public PlanWithProperties visitTableExecute(TableExecuteNode node, StreamPreferredProperties parentPreferences)
         {
-            return visitTableWriter(node, node.getPartitioningScheme(), node.getSource(), parentPreferences, node.getTarget());
+            return visitTableWriter(
+                    node,
+                    node.getPartitioningScheme(),
+                    node.getPreferredPartitioningScheme(),
+                    node.getSource(),
+                    parentPreferences,
+                    node.getTarget());
         }
 
         private PlanWithProperties visitTableWriter(
                 PlanNode node,
-                Optional<PartitioningScheme> partitioningSchemeOptional,
+                Optional<PartitioningScheme> partitioningScheme,
+                Optional<PartitioningScheme> preferredPartitionScheme,
                 PlanNode source,
                 StreamPreferredProperties parentPreferences,
                 WriterTarget writerTarget)
         {
-            return partitioningSchemeOptional
-                    .map(partitioningScheme -> visitPartitionedWriter(node, partitioningScheme, source, parentPreferences, writerTarget))
+            if (isTaskScaleWritersEnabled(session)
+                    && getTaskPartitionedWriterCount(session) > 1
```

Comment:
seems it would be cleaner to move this check to `visitScalePartitionedWriter` and add:
```
            if (getTaskPartitionedWriterCount(session) == 1) {
                return planAndEnforceChildren(node, singleStream(), defaultParallelism(session));
            }
```
within `visitScalePartitionedWriter` method.

It would be consistent with how `visitPartitionedWriter` works